src/Avalonia.Visuals/Rect.cs

```diff
@@ -421,11 +421,48 @@ public Rect Translate(Vector offset)
         }
 
         /// <summary>
-        /// Gets the union of two rectangles.
-        /// </summary>
-        /// <param name="rect">The other rectangle.</param>
-        /// <returns>The union.</returns>
-        public Rect Union(Rect rect)
+		/// Normalizes the rectangle so both the <see cref="Width"/> and <see cref="Height"/> are positive, without changing the location of the rectangle
+		/// </summary>
+        /// <returns>Normalized Rect</returns>
+		/// <remarks>
+		/// Empty rect will be return when Rect contains invalid values. Like NaN.
+		/// </remarks>
+		public Rect Normalize()
+        {
+            Rect rect = this;            
+
+            if(double.IsNaN(rect.Right) || double.IsNaN(rect.Bottom) || double.IsNaN(rect.X) || double.IsNaN(rect.Y) || double.IsNaN(Height) || double.IsNaN(Width))
+            {
+                return Rect.Empty;
```

Comment:
If we return Rect.Empty here shouldn't IsEmpty follow the same rules?

Reply:
Some checks could be redundant here. Bottom and Right are composed values. Is this true? `_x(NaN) + _width(123) == NaN`

Reply:
> If we return Rect.Empty here shouldn't IsEmpty follow the same rules?

I don't think so: this is normalizing a rect with invalid values to empty, it's not saying that a rect with invalid values _is_ empty: a rect with a positive width/height is definitely not an empty rect, although it's an invalid rect.